pkg/conversion/converter.go

```diff
@@ -0,0 +1,245 @@
+		structuralSchemas: structuralSchemas,
+		apiConversion:     apiConversion,
+	}, nil
+}
+
+type Converter struct {
+	crd               *apiextensionsv1.CustomResourceDefinition
+	compiledRules     map[string][]*CompiledRule
+	structuralSchemas map[string]*structuralschema.Structural
+	apiConversion     *apisv1alpha1.APIConversion
+}
+
+func (c *Converter) Convert(list *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
+	ctx := context.TODO()
+
+	convertedList := &unstructured.UnstructuredList{}
+	for i := range list.Items {
+		original := &list.Items[i]
+
+		converted, err := c.convert(ctx, original, targetGV)
+		if err != nil {
+			return nil, err
+		}
+
+		convertedList.Items = append(convertedList.Items, *converted)
+	}
+
+	return convertedList, nil
+}
+
+func (c *Converter) convert(ctx context.Context, original *unstructured.Unstructured, targetGV schema.GroupVersion) (*unstructured.Unstructured, error) {
+	converted := original.DeepCopy()
+	converted.SetAPIVersion(targetGV.String())
+
+	originalVersion := original.GetObjectKind().GroupVersionKind().Version
+	compiledRules := c.compiledRules[originalVersion]
+	for _, rule := range compiledRules {
+		value, err := evaluateRule(ctx, original, rule)
+		if err != nil {
+			return nil, fmt.Errorf("error converting: %w", err)
+		}
+
+		toPath := rule.ToPath
+		if toPath[0] == '.' {
+			toPath = toPath[1:]
+		}
+		fields := strings.Split(toPath, ".")
+
+		if err := unstructured.SetNestedField(converted.Object, value, fields...); err != nil {
+			return nil, fmt.Errorf("error setting destination field %q: %w", rule.ToPath, err)
+		}
+	}
+
+	annotations := converted.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	preserveAnnotation := annotations[targetGV.Version+apisv1alpha1.VersionPreservationAnnotationKeySuffix]
+	if preserveAnnotation != "" {
+		// If we're going to a version that already has the preserve annotation, restore from it
+		m := map[string]string{}
+		if err := json.Unmarshal([]byte(preserveAnnotation), &m); err != nil {
+			return nil, fmt.Errorf("error unmarshaling preserve annotation: %v", err)
+		}
+		if err := restoreFromPreserveMap(m, converted.Object); err != nil {
+			return nil, fmt.Errorf("error processing preserve map: %w", err)
+		}
+
+		// Don't save the preserve annotation
+		delete(annotations, targetGV.Version+apisv1alpha1.VersionPreservationAnnotationKeySuffix)
+		converted.SetAnnotations(annotations)
+	} else {
+		// Otherwise, store it
+		m, err := generatePreserveMap(original, c.apiConversion.Spec.Conversions)
+		if err != nil {
+			return nil, fmt.Errorf("error preserving fields: %w", err)
+		}
+
+		if len(m) > 0 {
+			encoded, err := json.Marshal(m)
+			if err != nil {
+				return nil, fmt.Errorf("error encoding preserve map: %v", err)
+			} else {
+				annotations[originalVersion+apisv1alpha1.VersionPreservationAnnotationKeySuffix] = string(encoded)
```

Comment:
From prior experience with our annotation-stuffing schemes - how can we be certain that targetGV.Version+apisv1alpha1.VersionPreservationAnnotationKeySuffix is a valid annotation key?

Reply:
It should be. If we find issues, we can adjust it.

Reply:
Doesn't this depend on the length of `originalVersion`? Is that bounded?

Reply:
If it's safer to hash, let's do that. @sttts ?

Reply:
Max length of originalVersion is 63 chars

Reply:
Will change to $prefix/$name - prefix can be 253, name 63 (https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set)